pkg/license/license.go

```diff
@@ -0,0 +1,329 @@
+	var (
+		preferredExt      = []string{"md", "markdown", "txt", "html"}
+		ignoredExt        = []string{"xml", "go", "gemspec", "spdx", "header"}
+		preferredExtRegex = regexp.MustCompile(`\.(?:` + regexp.QuoteMeta(preferredExt[0]) + `|` + regexp.QuoteMeta(preferredExt[1]) + `|` + regexp.QuoteMeta(preferredExt[2]) + `|` + regexp.QuoteMeta(preferredExt[3]) + `)$`)
+		anyExtRegex       = regexp.MustCompile(`(\.[^./]+$)`)
+		licenseRegex      = regexp.MustCompile(`(?i)(un)?licen[sc]e`)
+		copyingRegex      = regexp.MustCompile(`(?i)copy(ing|right)`)
+		oflRegex          = regexp.MustCompile(`(?i)ofl`)
+		patentsRegex      = regexp.MustCompile(`(?i)patents`)
+		filenameRegexes   = map[*regexp.Regexp]float64{
+			regexp.MustCompile(`(?i)^` + licenseRegex.String() + `$`):                                          1.00, // LICENSE
+			regexp.MustCompile(`(?i)^` + licenseRegex.String() + preferredExtRegex.String() + `$`):             0.95, // LICENSE.md
+			regexp.MustCompile(`(?i)^` + copyingRegex.String() + `$`):                                          0.90, // COPYING
+			regexp.MustCompile(`(?i)^` + copyingRegex.String() + preferredExtRegex.String() + `$`):             0.85, // COPYING.md
+			regexp.MustCompile(`(?i)^` + licenseRegex.String() + anyExtRegex.String() + `$`):                   0.80, // LICENSE.textile
+			regexp.MustCompile(`(?i)^` + copyingRegex.String() + anyExtRegex.String() + `$`):                   0.75, // COPYING.textile
+			regexp.MustCompile(`(?i)^` + licenseRegex.String() + `[-_][^.]*` + anyExtRegex.String() + `?$`):    0.70, // LICENSE-MIT
+			regexp.MustCompile(`(?i)^` + copyingRegex.String() + `[-_][^.]*` + anyExtRegex.String() + `?$`):    0.65, // COPYING-MIT
+			regexp.MustCompile(`(?i)^\w+[-_]` + licenseRegex.String() + `[^.]*` + anyExtRegex.String() + `?$`): 0.60, // MIT-LICENSE-MIT
+			regexp.MustCompile(`(?i)^\w+[-_]` + copyingRegex.String() + `[^.]*` + anyExtRegex.String() + `?$`): 0.55, // MIT-COPYING
+			regexp.MustCompile(`(?i)^` + oflRegex.String() + preferredExtRegex.String()):                       0.50, // OFL.md
+			regexp.MustCompile(`(?i)^` + oflRegex.String() + anyExtRegex.String()):                             0.45, // OFL.textile
+			regexp.MustCompile(`(?i)^` + oflRegex.String() + `$`):                                              0.40, // OFL
+			regexp.MustCompile(`(?i)^` + patentsRegex.String() + `$`):                                          0.35, // PATENTS
+			regexp.MustCompile(`(?i)^` + patentsRegex.String() + anyExtRegex.String() + `$`):                   0.30, // PATENTS.txt
+		}
+	)
+
+	var licenseFiles []LicenseFile
+	var ignore bool
+
+	err := fs.WalkDir(fsys, ".", func(filePath string, info fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			return nil
+		}
+
+		// Let's ignore all files in the melange-out/ directory, as it's not part of the source
+		if strings.Contains(filePath, "melange-out") {
```

Comment:
should we use `melangeOutputDirName`?

Reply:
As `melangeOutputDirName` is defined in the build package, and the build package is using the license package, we'd have to move it somewhere out to not introduce a circular dependency (and make it exported). I think this is a valid thing to do, but I might tackle it a bit later.